### DISCUSSION TITLE
beginner question about `st`

### DISCUSSION THREAD

**Discussion by MangoIV** [koka-lang/koka#554] [Q&A] [answered]
I was playing around with koka a bit and encountered something that surprised me, and it's probably due to me not understanding syntax: 

```
fun i(x : int, r : ref<h, int>) : <st<h>> () { 
  modify(r) fn(i)
    i + x
  ()
}

fun j() : div int { 
  run {
    var r := ref(0)
    i(5, r)
    !r
  }
}
```

why does `j()` return `0` instead of `5`, what is going wrong there? 

I was expecting this to work like the following program in Haskell: 

```hs
i :: Int -> STRef s Int -> ST s ()
i x ref = do
  modifySTRef ref \i -> x + i

j :: Int
j = runST do
  r <- newSTRef 0
  i 5 r
  readSTRef r
```

**Comment by TimWhiting** [comment, marked as answer]
Ah yes this confused me at first too. The return value of modify is the return value of the lambda you pass it, the return value of modify is not used to set the ref. Inside the lambda `i` refers to a local variable (you can use `i := i + x` to get what you want). It's closer in behavior and semantics to Haskell's modifyMVar without the need to return a tuple.

**Reply by MangoIV** [reply]
so `i` is actually the mutable reference and not the value stored in the mutable reference?

**Reply by MangoIV** [reply]
this is really confusing :D 

Thank you for clarifying!
